django/contrib/auth/forms.py

```diff
@@ -212,14 +219,32 @@ def clean(self):
         password = self.cleaned_data.get('password')
 
         if username is not None and password:
+            self.pre_authenticate(username)
             self.user_cache = authenticate(self.request, username=username, password=password)
             if self.user_cache is None:
+                # Cache failed login with delay.
```

Comment:
We (and probably others) have existing solutions for this issue, and so would prefer to turn this off, both to avoid conflicts with existing policies and to avoid the slight cache overhead. So I suggest skipping the checks if `DELAY_AFTER_FAILED_LOGIN` is set to 0.

Reply:
I don't use login rate limiting myself at the moment, but I can see that people might want to use more sophisticated schemes (e.g. exponential delays). So I think it would be good to allow people to opt-out of this feature.